iree/turbine/kernel/wave/scheduling/loop_reconstruction.py

```diff
@@ -99,39 +104,56 @@ def add_nodes_by_schedule(
             # Set the index for the new node by substituting the induction variable
             # for the current iteration.
             new_node.index = node.index
-            for dim in new_node.index:
-                new_node.index[dim] = new_node.index[dim].subs(
-                    {induction_variable: current_induction_variables[iteration]}
-                )
+            if new_node.index:
+                for dim in new_node.index:
+                    new_node.index[dim] = new_node.index[dim].subs(
+                        {induction_variable: current_induction_variables[iteration]}
+                    )
+            if custom_node.expanded_dims:
+                new_node.expanded_dims = custom_node.expanded_dims
             # Add scheduling parameters for debugging.
             new_node.scheduling_parameters = node.scheduling_parameters
             # Update the rotating registers and argument context for the current node (if applicable).
+            old_node = None
             if node in rotating_registers:
                 rotating_registers[node].append(new_node.fx_node)
-                rotating_registers[node].popleft()
+                old_node = rotating_registers[node].popleft()
                 # If draining, then override the rotating registers and update the argument context.
                 if fill_or_drain:
                     for next_stage in range(stage + 1, len(stages)):
                         arg_context[(iteration, next_stage, node)] = new_node.fx_node
 
-            # Update the init args in the argument context whenever a result is computed.
+            # Update the iter and init args in the argument context whenever a result is computed.
             if node in arg_context.results:
+                iter_arg = arg_context.result_to_iter_arg[node]
+                logger.debug(
+                    f"Updating result: {node} -> {iter_arg} to {new_node.fx_node}."
+                )
+                arg_context.map_arg_all_after_iteration(
+                    iter_arg,
+                    new_node.fx_node,
+                    iteration,
+                )
                 if (
                     pipelining_stage == PipelineStage.KERNEL
-                    or pipelining_stage == PipelineStage.EPILOGUE
+                    or pipelining_stage == PipelineStage.PROLOGUE
```

Comment:
can you explain this change little bit more?

Reply:
sure, will also add a comment. 
                # In situations where we have an iter_arg as a rotating register,
                # we also have the output as a rotating register. So when we
                # are updating the output, we update the iter_arg as well with the
                # old value of the output rotating register. Consider this example:
                # Say we have the following:
                #
                # Stage 0:
                # iter_arg0
                #
                #
                # output = compute(...) -> here we update iter_arg0 to have the output value
                #                          for the next stage, so that it gets picked up in stage1.
                #
                # Stage 1:
                # b = use(iter_arg0)